### ISSUE TITLE
CPU bf16/fp16 matmul via sbgemm_/shgemm_ throws on legal shapes: `int c_size = n_ * m_` overflows for outputs >= 2^31 elements

### ISSUE THREAD

**Issue by BwL1289** [pytorch/pytorch#191083]
### 🐛 Describe the bug

In `aten/src/ATen/native/CPUBlas.cpp`, the bf16 (`sbgemm_`) and fp16 (`shgemm_`) external-BLAS copy-back paths compute the float staging buffer size in 32-bit `int`:

```cpp
int c_size = n_ * m_;
// C matrix in OpenBLAS sbgemm are of type "float" so we have to convert, copy and copy back.
std::vector<float> float_v(c_size, 0.0f);
```

(`https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/CPUBlas.cpp` — the `BLAS_HAS_SBGEMM` block, ~line 373, and its `BLAS_HAS_SHGEMM` twin, ~line 434.)

`use_blas_gemm` guards each of `m, n, k, lda, ldb, ldc <= INT_MAX` individually, but not the **product** `m*n`. For a legal bf16 matmul whose output has ≥ 2³¹ elements — e.g. 46,341 × 46,341 (each dim tiny relative to INT_MAX; product 2,147,488,281 > 2³¹−1) — `n_ * m_` overflows signed int (UB; wraps negative in practice), and the `std::vector` constructor throws.

Reproduction (CPU build where the BLAS provides `sbgemm_`, i.e. `BLAS_HAS_SBGEMM` is defined — e.g. OpenBLAS with BUILD_BFLOAT16):

```python
import torch
m = n = 46341
k = 64
a = torch.randn(m, k, dtype=torch.bfloat16)
b = torch.randn(k, n, dtype=torch.bfloat16)
c = a @ b   # RuntimeError (std::length_error "vector" from the negative size)
```

Measured on linux-aarch64 (Neoverse V2), OpenBLAS 0.3.33.dev: 46,340² (product just under 2³¹) completes and matches an fp32 reference; 46,341² throws immediately.

### Expected behavior

Either compute `c_size` in `int64_t` (a 2.1G-element float staging buffer is a legal ~8.6 GB allocation, and the surrounding copy loops already index with `int64_t` loop variables), or route these shapes to the internal gemm fallback the way the per-dimension guard already does.

### Versions

PyTorch 2.14 dev (main-lineage, ≈`6a641f6777594fcd2f34ea32f7ee2c0cdaa55776`), linux-aarch64 (Neoverse V2), OpenBLAS 0.3.33.dev LP64 with `sbgemm_` exported. Related to the wider 32-bit-LAPACK/BLAS discussion in #51720 / #190509 (this one is torch-side arithmetic, not LAPACK-side).

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @jianyuh @nikitaved @mruberry @walterddr @Lezcano @snadampal @milpuz01 @nikhil-arm @fadara01 @robert-hardwick @nWEIdia

**Comment by Diacod-I** [comment]
Would love to take up this issue.

**Comment by aditew01** [comment]
Thanks for raising this issue @BwL1289

**Comment by Diacod-I** [comment]
@BwL1289 
Update: I have created a PR ( #191096 ) to fix this issue. Feel free to reach out if you think I need to change anything in this. Thank you for bringing up this issue!
